optaplanner-core/src/main/java/org/optaplanner/core/impl/score/holder/AbstractScoreHolder.java

```diff
@@ -106,20 +118,23 @@ public void configureConstraintWeight(Rule rule, Score_ constraintWeight) {
      * @param kcontext The rule for which to register the match.
      * @param constraintUndoListener The operation to run to undo the match.
      * @param scoreSupplier The score change to be undone when constraint justification enabled.
+     * @param justifications the primary arguments(s) that the CS penalizes/rewards, empty when not called by CS-D
      */
-    protected void registerConstraintMatch(RuleContext kcontext,
-            final Runnable constraintUndoListener, Supplier<Score_> scoreSupplier) {
-        AgendaItem<?> agendaItem = (AgendaItem) kcontext.getMatch();
+    protected void registerConstraintMatch(RuleContext kcontext, Runnable constraintUndoListener,
+            Supplier<Score_> scoreSupplier, Object... justifications) {
+        AgendaItem<?> agendaItem = (AgendaItem<?>) kcontext.getMatch();
         ConstraintActivationUnMatchListener constraintActivationUnMatchListener = new ConstraintActivationUnMatchListener(
                 constraintUndoListener);
         agendaItem.setCallback(constraintActivationUnMatchListener);
         if (constraintMatchEnabled) {
-            List<Object> justificationList = extractJustificationList(kcontext);
+            // If we have justifications coming from CS-D directly, use them exclusively.
+            List<Object> completeJustificationList =
+                    justifications.length == 0 ? extractJustificationList(kcontext) : Arrays.asList(justifications);
```

Comment:
This change will make constraint matches in CS-D perform about a bazillion times faster.